crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/meta_net_task.rs

```diff
@@ -0,0 +1,1157 @@
+                con: mk_test_con(),
+            })
+            .await
+            .unwrap();
+
+        tokio::time::timeout(Duration::from_millis(100), async {
+            while !internal_stub.connections.read().is_empty() {
+                tokio::time::sleep(Duration::from_millis(1)).await;
+            }
+        })
+        .await
+        .expect("Timed out waiting for connection to be removed");
+
+        assert_eq!(0, internal_stub.connections.read().len());
+    }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn handle_disconnect_stops_task_if_internal_sender_closes() {
+        let (mut ep_evt_send, _, internal_sender, _, _, meta_net_task_finished) = setup().await;
+
+        internal_sender
+            .ghost_actor_shutdown_immediate()
+            .await
+            .unwrap();
+
+        ep_evt_send
+            .send(MetaNetEvt::Disconnected {
+                remote_url: "".to_string(),
+                con: mk_test_con(),
+            })
+            .await
+            .unwrap();
+
+        tokio::time::timeout(Duration::from_millis(1000), async {
+            while !meta_net_task_finished.load(Ordering::Acquire) {
+                tokio::time::sleep(Duration::from_millis(1)).await;
+            }
+        })
+        .await
+        .expect("Timed out waiting for task to shut down");
+
+        assert!(meta_net_task_finished.load(Ordering::Acquire));
+    }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn make_request_while_blocked() {
+        let (mut ep_evt_send, _, _, _, host_stub, _) = setup().await;
+
+        host_stub
+            .block(Block::new(
+                BlockTarget::Node(test_node_id(1), NodeBlockReason::DOS),
+                InclusiveTimestampInterval::try_new(
+                    Timestamp::now(),
+                    Timestamp::now()
+                        .checked_add(&Duration::from_secs(10))
+                        .unwrap(),
+                )
+                .unwrap(),
+            ))
+            .await
+            .unwrap();
+
+        let con = mk_test_con_with_id(1);
+        let con_state = get_con_state(&con);
+
+        ep_evt_send
+            .send(MetaNetEvt::Request {
+                remote_url: "".to_string(),
+                con: con,
+                data: wire::Wire::Call(wire::Call {
+                    space: test_space(1),
+                    to_agent: test_agent(2),
+                    data: WireData(vec![]),
+                }),
+                respond: Box::new(|_| async move { () }.boxed().into()),
+            })
+            .await
+            .unwrap();
+
+        tokio::time::timeout(Duration::from_millis(1000), async {
+            while !con_state.read().closed {
+                tokio::time::sleep(Duration::from_millis(1)).await;
+            }
+        })
+        .await
+        .expect("Timed out waiting for the connection to be closed");
+
+        assert!(con_state.read().closed);
```

Comment:
In this case, the connection gets closed but we don't tell the host it's gone. Is there a mechanism that will sweep for closed connections? I haven't looked too closely yet but this feels incomplete as an error handling mechanism

Reply:
Not sure what "host" refers to in your question. If the remote is blocked for any reason, it would be incorrect to engage in any further communication with them. As far as conductor is concerned, it *should* already get a blocked error when making an outgoing request, but this is an incoming request.

Reply:
Host is my mistake, it's the internal sender that gets connect/disconnect messages which doesn't actually send messages to the host in this case. A better description of what I'm querying is: When we locally force-close a connection from a blocked node, we don't trigger a disconnected event. That means that consumers of this task's events won't know that connection is gone. This actually only affects metrics at the moment as far as I can see, but even so this is a state management gap.

Reply:
I've put in a TODO to take a look at this